### ISSUE TITLE
Enforce X amount of concurrent operations

### ISSUE THREAD

**Issue by yaronel** [AppsFlyer/aerospike-clj#25]
**Is your feature request related to a problem? Please describe.**

When you set the `maxCommandsInQueue` option of the event policy, Aerospike starts throwing exceptions once you try to execute another command and the queue is full.
Aerospike say in their docs that they expect the user to implement a mechanism for handling this issue. It would be nice if the library used a semaphore and handled it for the user.

**Describe the solution you'd like**

We can use a Semaphore with permits number equal to the number of `maxCommandsInQueue`. 
Here's a simple wrapper around Semaphore in clojure:
```clojure
(ns af-message-validation.utils.semaphore
  (:import [java.util.concurrent Semaphore]))

(defprotocol PermitProvider
  (acquire [this])
  (release [this])
  (available [this]))

(defrecord SemaphoreCLJ [^Semaphore impl]
  PermitProvider
  (acquire [_]
    (.acquire ^Semaphore impl))
  (release [_]
    (.release ^Semaphore impl))
  (available [_]
    (.availablePermits ^Semaphore impl)))

(defn create-semaphore
  ([permits]
   (create-semaphore permits false))
  ([permits fairness]
   (->SemaphoreCLJ (Semaphore. permits fairness))))
```

And then we can acquire and release the permit before scheduling the command with a macro:
```clojure
(defmacro with-permit
  "Runs the body after acquiring a permit. The permit is released after the body returns."
  [& body]
  `(do
     (get-permit)
     (let [res# ~@body]
       (release-permit)
       res#)))
```

**Describe alternatives you've considered**

The user can do the same thing in the code, but it seems like something that should be in the library.

**Additional context**

**Comment by barkanido** [comment]
Okay, l think we should add it as an option. After all, some users might be surprised that it changes a fundamental behavior if the Java library (although I agree the default implementation is kind of problematic).

**Comment by eranharel** [comment]
What is the benefit of adding concurrency control on top of the existing concurrency control?
What you really want to implement is a back-off mechanism so that the users don't have to implement it over and over again.

Apart from that, the use of the semaphore will break the non-blocking API (makes it blocking)

**Comment by yaronel** [comment]
Maybe a semaphore is not the right choice. 
But do we agree that for many cases the desired behaviour once an attempt to add a command to a queue which is full, is for the thread to wait until a slot is available, rather than throw an exception?

**Comment by barkanido** [comment]
@eranharel this is not adding concurrency control over an existing one. This is trying to offer a more convenient behavior over a queue. The current mechanism of a queue that fails (with an exception) when it is full just makes implementing backoff (probably) less convenient. I believe that what @yaronel offers is to make it easier to do so.
Also, in my opinion, blocking when the queue of async command is full is exactly backing off. This implies the calling code that it just cannot push more commands to an already too loaded cluster.
Some users will prefer catching an exception and do some time-based backoff, others might want just wait until there is room in the queue again. This is just an optional addition.

**Comment by eranharel** [comment]
Blocking is one way of backing off. Another way is non blocking retries with a growing backoff delay.
What I said above is that you already get the signal you need in order to respond to the load. You don't need to add another mechanism, and if you do add semaphores, than this client will no longer remain non-blocking

**Comment by barkanido** [comment]
I am fine with that as long as that is an optional behavior that a user chooses to enable. This is a very simple and cheap mechanism that although would "officially" make this client blocking, will allow simple backoff with a most-of-the-time non-blocking fast behavior.
So, in practice, when the cluster is fast enough and healthy, and the client(s) are not overloading either the cluster or the concurrency limits of the event-loop for too long, the user can expect the client to be non-blocking until these assumptions change.
I can agree that some users will not want it and want a more sophisticated time-based back-off mechanism.

**Comment by barkanido** [comment]
Another option is to expose a configuration hook for the user code. If there is `maxCommandsInQueue` configured and a new switch called `blockOnQueueFull` is on, then the client will attempt to acquire a semaphore of exactly `maxCommandsInQueue + maxCommandsInProcess` permits and will release it as soon as the command listener is called.

Pros:
- should be performant
- play nicely with the client native configuration

Cons:
- each API call should be changed to support this. Currently, since `register-events` is built on the underlying client listeners, it only supports chaining stuff upon completion, so the acquire part cannot be a part of it.
- this new flag is not a part of a `Policy` which is the least surprising place to put configuration on the Java client.

**Comment by yaronel** [comment]
I am not sure we can just add a semaphore on entry / exit.

Let's say we have a queue of 1. We take a permit when submitting a task and release it `onSuccess`. The `onSuccess` callback is running on the event loop thread. 
- If we release the permit before calling the user's callback, then won't we be allowing another task to be added to the full queue? 
- If we release it after the user's callback, then we are occupying the event loop thread with user code which slows down the request rate, and will cause a deadlock if the user tries to make another db call on that thread.

I think we should separate the user's callback from the event loop thread. We can acquire a permit when submitting a task, but the user's callback should not be executed on the event loop thread. When the `onSuccess` event is triggered we should put the response on an unbounded queue. A worker thread will take responses from the queue and call the user's callback. 
The unbounded queue allows for bursts of requests. In cases where the request rate is constantly faster than the response rate the queue will just fill up and a profiler will make it clear to the user that their implementation needs tweaking.

I am not sure what will be the implications of running another dedicated worker thread though.

**Comment by barkanido** [comment]
To be clear, my definition of call-back is the `*Listener` interface (`OnSuccess`+`OnFalure`). So there is no such thing as a 'user callback' technically because it all happens on the event loop threads (even if the user supplied the code and context for it).

> If we release the permit before calling the user's callback, then won't we be allowing another task to be added to the full queue?

All callbacks are done in the event-loop threads. This is by definition. Also, if all async API's properly respect the discipline of trying to acquire the lock then this will work.

> If we release it after the user's callback, then we are occupying the event loop thread with user code which slows down the request rate, and will cause a deadlock if the user tries to make another DB call on that thread.

We cannot release outside the call-back since the call-back is the only place we know that the command returned.

> I think we should separate the user's callback from the event loop thread.

I don't quite agree. First, this breaks the API. Second, it complicates the runtime model significantly. Third, creating an unbounded queue and counting on profiling is considered an antipattern because it leads to nasty and unexpected memory leaks.

The only problem as I see in acquiring the semaphore before issuing the command is that the async API can become a blocking one when the client is overloaded.

**Comment by barkanido** [comment]
added #32 draft to support this. tell me what you think.

**Comment by chorin1** [comment]
Aerospike's java async API assumes the user is taking care of not pushing too many commands to the queue: 
> These EventPolicy defaults assume the user will apply an external throttle to mitigate event loop overload. The example [AsyncTest](https://www.aerospike.com/docs/client/java/usage/async/) performs this throttle by seeding the event loop with N commands and running exactly one more command when a command completes.

I think the Clojure client shouldn't act any different, otherwise as @eranharel said - this is no more an async client.

I guess adding this as an optional feature would prove useful - 
whether it will be implemented by Semaphores or the more idiomatic (but probably less performant) using channels.

**Comment by barkanido** [comment]
> I guess adding this as an optional feature would prove useful -
> whether it will be implemented by Semaphores or the more idiomatic (but probably less performant) using channels.

This patch **does** adds this functionality as optional.

**Comment by barkanido** [comment]
Closing this one as it will be solved outside of this project (#32 )
